### PR TITLE
fix compatibility with SCOOP 0.7.x

### PR DESCRIPTION
**Pull request by boegel**
This is required to make `myscoop` work on top of SCOOP 0.7.x, due to API changes.

Running the `squares.py` example locally now yields:

```
$ myscoop --sched=local --scoop-module=squares 10000
[2015-09-03 13:05:01,434] launcher  INFO    SCOOP 0.7 1.1 on linux2 using Python 2.7.10 (default, Sep  2 2015, 19:01:27) [GCC Intel(R) C++ gcc 4.9 mode], API: 1013
[2015-09-03 13:05:01,434] launcher  INFO    Deploying 16 worker(s) over 1 host(s).
[2015-09-03 13:05:01,434] launcher  INFO    Worker distribution: 
[2015-09-03 13:05:01,434] launcher  INFO       localhost:   15 + origin
2015-09-03 13:05:04,275 ERROR      runpy.CpuSetT   MainThread  set_cpus: length cpu list 1024 is larger then cpusetsize 64. Truncating to cpusetsize
2015-09-03 13:05:04,275 ERROR      runpy.CpuSetT   MainThread  set_cpus: length cpu list 1024 is larger then cpusetsize 64. Truncating to cpusetsize
2015-09-03 13:05:04,275 ERROR      runpy.CpuSetT   MainThread  set_cpus: length cpu list 1024 is larger then cpusetsize 64. Truncating to cpusetsize
2015-09-03 13:05:04,309 ERROR      runpy.CpuSetT   MainThread  set_cpus: length cpu list 1024 is larger then cpusetsize 64. Truncating to cpusetsize
2015-09-03 13:05:04,321 ERROR      runpy.CpuSetT   MainThread  set_cpus: length cpu list 1024 is larger then cpusetsize 64. Truncating to cpusetsize
2015-09-03 13:05:04,458 ERROR      runpy.CpuSetT   MainThread  set_cpus: length cpu list 1024 is larger then cpusetsize 64. Truncating to cpusetsize
2015-09-03 13:05:04,525 ERROR      runpy.CpuSetT   MainThread  set_cpus: length cpu list 1024 is larger then cpusetsize 64. Truncating to cpusetsize
2015-09-03 13:05:04,529 ERROR      runpy.CpuSetT   MainThread  set_cpus: length cpu list 1024 is larger then cpusetsize 64. Truncating to cpusetsize
2015-09-03 13:05:04,530 ERROR      runpy.CpuSetT   MainThread  set_cpus: length cpu list 1024 is larger then cpusetsize 64. Truncating to cpusetsize
2015-09-03 13:05:04,537 ERROR      runpy.CpuSetT   MainThread  set_cpus: length cpu list 1024 is larger then cpusetsize 64. Truncating to cpusetsize
2015-09-03 13:05:04,538 ERROR      runpy.CpuSetT   MainThread  set_cpus: length cpu list 1024 is larger then cpusetsize 64. Truncating to cpusetsize
2015-09-03 13:05:04,538 ERROR      runpy.CpuSetT   MainThread  set_cpus: length cpu list 1024 is larger then cpusetsize 64. Truncating to cpusetsize
2015-09-03 13:05:04,540 ERROR      runpy.CpuSetT   MainThread  set_cpus: length cpu list 1024 is larger then cpusetsize 64. Truncating to cpusetsize
2015-09-03 13:05:04,542 ERROR      runpy.CpuSetT   MainThread  set_cpus: length cpu list 1024 is larger then cpusetsize 64. Truncating to cpusetsize
2015-09-03 13:05:04,542 ERROR      runpy.CpuSetT   MainThread  set_cpus: length cpu list 1024 is larger then cpusetsize 64. Truncating to cpusetsize
2015-09-03 13:05:04,542 ERROR      runpy.CpuSetT   MainThread  set_cpus: length cpu list 1024 is larger then cpusetsize 64. Truncating to cpusetsize
First 10000 squares: [0, 1, 4, 9, 16, 25, 36, 49, ...
[2015-09-03 13:05:07,430] launcher  (127.0.0.1:36046) INFO    Root process is done.
[2015-09-03 13:05:07,431] launcher  (127.0.0.1:36046) INFO    Finished cleaning spawned subprocesses.
```

I'm not sure what the `ERROR` is about (yet).

@stdweird: please review

### CODE REVIEW

File: lib/vsc/mympirun/scoop/myscoop.py [hunk 1]
```diff
@@ -337,27 +347,56 @@ def scoop_run(self):
...
+                # custom
+                self.scoop_freeorigin,
+                self.scoop_processcontrol,
+                self.scoop_affinity,
+                vars_to_pass,
+            ]
+
         self.log.debug("scoop_run: scoop_app class %s args %s" % (self.SCOOP_APP.__name__, scoop_app_args))
 
         scoop_app = self.SCOOP_APP(*scoop_app_args)
```
Comment (JensTimmerman) [inline lib/vsc/mympirun/scoop/myscoop.py]: this is bound to break again in the future, can't we pass named arguments? and a nice error if it seems like the api changed?

**Comment by JensTimmerman** [comment]
This needs tests.

**Comment by boegel** [comment]
I already spent quite a bit of time on this, I can't really spend more time on it right now, too much other stuff is piling up.

I do agree that various things could be handled better here, and tests would be nice too indeed.

I'll go ahead and install an older version so the user can continue, this will have to be picked up later again.

**Comment by JensTimmerman** [comment]
fixing things correctly will save you time in the long run.

**Comment by boegel** [comment]
I agree, but I don't have time to fix these things correctly, I already spent way more time on this than I was planning to (but I didn't want to let that work go to waste, so I continued until I got things working, which is the 1st step).

File: lib/vsc/mympirun/scoop/myscoop.py [hunk 2]
```diff
@@ -337,27 +347,56 @@ def scoop_run(self):
...
-                          "other",  # TODO check utils.getEnv(),
-                          self.scoop_profile,
-                          self.scoop_pythonpath[0],
-                          # custom
-                          self.scoop_freeorigin,
-                          self.scoop_processcontrol,
-                          self.scoop_affinity,
-                          vars_to_pass,
-                          ]
+        if LooseVersion(SCOOP_VERSION) >= LooseVersion('0.7'):
```
Comment (stdweird) [inline lib/vsc/mympirun/scoop/myscoop.py:378]: can you set the old list and insert / change things for version 0.7. this is way too much copy/paste

**Comment by boegel** [comment]
Remarks fixed, please reconsider.

Jenkins doesn't agree because no test suite is available yet for this, but I'd prefer not to dive into that now (would be a lot of extra effort).

**Comment by JensTimmerman** [comment]
jenkins: test this please

**Comment by JensTimmerman** [comment]
jenkins disagrees because vsc-mympirun 3.4.0 is not released yet

**Comment by boegel** [comment]
@JensTimmerman: you mean, on PyPI? vsc-processcontrol isn't there either yet

**Comment by JensTimmerman** [comment]
yes, on pypi, correctly building it on jenkins would already fix these tests, but it's still using an old broken shared_setup, see https://github.com/hpcugent/vsc-mympirun/issues/43

**Comment by boegel** [comment]
last commit includes some more fixes related to SCOOP 0.7.x

will need some cleanup before it can be merged, but this seems to work (cfr. OTRS ticket 10250047)

**Comment by boegel** [comment]
The changes in here should be accompanied by patches to SCOOP, which are required to start all workers through SSH and make sure that the worker environment is set up properly for _all_ workers.

``` diff

--- scoop-0.7.1.1/scoop/launch/workerLaunch.py.orig 2016-02-21 20:41:47.000000000 +0100
+++ scoop-0.7.1.1/scoop/launch/workerLaunch.py  2016-02-21 20:43:00.000000000 +0100
@@ -109,7 +109,8 @@
             broker = worker.brokerHostname

         # If host is not localhost, echo group process
-        if not self.isLocal() and workerID == 0:
+        #if not self.isLocal() and workerID == 0:
+        if workerID == 0:
             c.append("--echoGroup ")

         if worker.nice is not None:
@@ -145,7 +146,7 @@
         # It replaces simple quotation marks with \\\" which gets evaluated to
         # \" by the second shell which prints it out as a double quote.
         if worker.args:
-            if self.isLocal():
+            if False: #self.isLocal():
                 # If local, no shell is used so no escaping needed
                 c.extend([
                     '{0}'.format(a) for a in worker.args
@@ -161,8 +162,8 @@
         worker = self.workersArguments[workerID]

         c = []
-        if not self.isLocal():
-            c.extend(self._WorkerCommand_environment(worker))
+        #if not self.isLocal():
+        c.extend(self._WorkerCommand_environment(worker))

         c.extend(self._WorkerCommand_bootstrap(worker))
         c.extend(self._WorkerCommand_options(worker, workerID))
@@ -188,7 +189,7 @@

     def launch(self, tunnelPorts=None, stdPipe=False):
         """Launch every worker assigned on this host."""
-        if self.isLocal():
+        if False: #self.isLocal():
             # Launching local workers
             for workerID, workerToLaunch in enumerate(self.workersArguments):
                 # Launch one per subprocess
--- scoop-0.7.1.1/scoop/launcher.py.orig    2016-02-21 21:07:01.000000000 +0100
+++ scoop-0.7.1.1/scoop/launcher.py 2016-02-22 16:10:24.000000000 +0100
@@ -299,7 +299,7 @@
                 (self.brokers[0].brokerPort,
                  self.brokers[0].infoPort)
                     if self.tunnel else None,
-                stdPipe=not self.workers[-1].isLocal(),
+                stdPipe=True, #not self.workers[-1].isLocal(),
             )
             if self.workersLeft <= 0:
                 # We've launched every worker we needed, so let's exit the loop
@@ -307,7 +307,7 @@
                 break

         # Wait for the root program
-        if self.workers[-1].isLocal():
+        if False: #self.workers[-1].isLocal():
             self.errors = self.workers[-1].subprocesses[-1].wait()
         else:
             # Process stdout first, then the whole stderr at the end
```

**Comment by boegel** [comment]
This is a patch to SCOOP that works a lot better (not sure if the `raise` thing is still needed, that was partially for debugging)

``` diff
diff -ru scoop-0.7.1.1.orig/scoop/encapsulation.py scoop-0.7.1.1/scoop/encapsulation.py
--- scoop-0.7.1.1.orig/scoop/encapsulation.py   2015-08-04 16:27:22.000000000 +0200
+++ scoop-0.7.1.1/scoop/encapsulation.py    2016-02-23 10:50:46.000000000 +0100
@@ -18,6 +18,7 @@
 import tempfile
 import types
 import os
+import sys
 from inspect import ismodule
 from functools import partial
 try:
@@ -161,7 +162,11 @@
 def pickleFileLike(file_):
     position = file_.tell()
     file_.seek(0)
-    data = file_.read()
+    try:
+        data = file_.read()
+    except IOError as err:
+        sys.stderr.write("ERROR: Failed to read from %s (%s) in pickleFileLike: %s\n" % (file_, getattr(file_, 'name', None), err))
+        raise pickle.PicklingError(err)
     file_.seek(position)
     return unpickleFileLike, (position, data)

diff -ru scoop-0.7.1.1.orig/scoop/launch/workerLaunch.py scoop-0.7.1.1/scoop/launch/workerLaunch.py
--- scoop-0.7.1.1.orig/scoop/launch/workerLaunch.py 2015-08-21 15:53:21.000000000 +0200
+++ scoop-0.7.1.1/scoop/launch/workerLaunch.py  2016-02-23 08:49:05.000000000 +0100
@@ -161,8 +161,8 @@
         worker = self.workersArguments[workerID]

         c = []
-        if not self.isLocal():
-            c.extend(self._WorkerCommand_environment(worker))
+        #if not self.isLocal():
+        c.extend(self._WorkerCommand_environment(worker))

         c.extend(self._WorkerCommand_bootstrap(worker))
         c.extend(self._WorkerCommand_options(worker, workerID))
@@ -193,8 +193,10 @@
             for workerID, workerToLaunch in enumerate(self.workersArguments):
                 # Launch one per subprocess
                 c = self._getWorkerCommandList(workerID)
+                cmd = 'bash -i -c "%s"' % ' '.join(c).replace('"', '\\\"')
+                print "cmd: %s" % cmd
                 self.subprocesses.append(
-                    subprocess.Popen(c)
+                    subprocess.Popen(cmd, shell=True)
                 )
         else:
             # Launching remotely
```

**Comment by stdweird** [comment]
@boegel in your patch, why do you start a bash session within the `shell=True`?

**Comment by boegel** [comment]
@stdweird: I'm not 100% all of this is needed, but it works ;)

Should be revisited and cleaned up, this is what I ended up with, after a lot of trial-and-error...

**Comment by boegel** [comment]
@stdweird: well, this explains it:

```
>>> import subprocess
>>> subprocess.Popen("bash -i -c 'module list'")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 710, in __init__
    errread, errwrite)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 1335, in _execute_child
    raise child_exception
OSError: [Errno 2] No such file or directory
>>> subprocess.Popen("bash -i -c 'module list'", shell=True)
<subprocess.Popen object at 0x10448b2d0>
>>> No modules loaded



```

**Comment by stdweird** [comment]
@boegel my point is that you don't need a `bash -c` when you have `shell=True`

**Comment by boegel** [comment]
@stdweird: I do, because I need an interactive shell (to make sure the `module` command is defined)

**Comment by boegel** [comment]
@stdweird:

```
>>> subprocess.Popen('module list', shell=True)
<subprocess.Popen object at 0x102e9f250>
>>> /bin/sh: module: command not found
```

**Comment by stdweird** [comment]
ah yes, can you add a comment to the patch?

**Comment by boegel** [comment]
@stdweird: I think we better figure out how we're going to handle this...

Merge this PR in vsc-mympirun-scoop, and add the required SCOOP patch somewhere in here too?

**Comment by stdweird** [comment]
@boegel you can add the patch here as reference, but patch for scoop should either be upstream or part of the easyconfig required for myscoop

File: lib/vsc/mympirun/scoop/myscoop.py [hunk 3]
```diff
@@ -63,6 +70,17 @@ class MyHost(Host):
                                       'variables', 'load_modules']
                                      )
 
+    def __init__(self, *args, **kwargs):
+        """Constructor; add logger if none is available"""
+        super(MyHost, self).__init__(*args, **kwargs)
+        if not hasattr(self, 'log'):
+            self.log = getLogger('MyHost')
+
+    #def isLocal(self):
```
Comment (stdweird) [inline lib/vsc/mympirun/scoop/myscoop.py:79]: remove if not required

File: lib/vsc/mympirun/scoop/myscoop.py [hunk 4]
```diff
@@ -127,15 +148,17 @@ def _WorkerCommand_options(self, worker, workerId):
...
-        self.processcontrol = args.pop()
-        self.freeorigin = args.pop()
-        super(MyScoopApp, self).__init__(*args)
+        self.load_modules = kwargs.pop('load_modules')
+        self.variables_to_pass = kwargs.pop('variables_to_pass')
+        self.affinity = kwargs.pop('affinity')
+        self.processcontrol = kwargs.pop('processcontrol')
+        self.freeorigin = kwargs.pop('freeorigin')
+        try:
+	        super(MyScoopApp, self).__init__(*args, **kwargs)
```
Comment (stdweird) [inline lib/vsc/mympirun/scoop/myscoop.py:159]: that's a lot of indent

File: lib/vsc/mympirun/scoop/myscoop.py [hunk 5]
```diff
@@ -337,35 +360,53 @@ def scoop_run(self):
...
-                          self.scoop_freeorigin,
-                          self.scoop_processcontrol,
-                          self.scoop_affinity,
-                          vars_to_pass,
-                          self.scoop_load_modules,
-                          ]
+        scoop_app_args = {
+            'arguments': self.scoop_args,
+            'debug': self.scoop_debug,
+            'env': "other", # TODO check utils.getEnv(),
```
Comment (stdweird) [inline lib/vsc/mympirun/scoop/myscoop.py:366]: open isse for the TODO

File: lib/vsc/mympirun/scoop/myscoop.py [hunk 6]
```diff
@@ -337,35 +360,53 @@ def scoop_run(self):
...
         self.log.debug("scoop_run: scoop_app class %s args %s" % (self.SCOOP_APP.__name__, scoop_app_args))
 
-        scoop_app = self.SCOOP_APP(*scoop_app_args)
+        scoop_app = self.SCOOP_APP(**scoop_app_args)
         try:
             root_task_ec = scoop_app.run()
             self.log.debug("scoop_run exited with exitcode %s" % root_task_ec)
         except Exception as e:
-            self.log.exception('scoop_run: error while launching SCOOP subprocesses: {0}'.format(str(e)))
+            print "Exception: %s" % e
```
Comment (stdweird) [inline lib/vsc/mympirun/scoop/myscoop.py:408]: is the print still needed?

**Comment by boegel** [comment]
This is still important, at some point, I'll need to find/make time for this and finish it up...

**Comment by stdweird** [comment]
that would be nice

**Pull request closed without merging**